### PR TITLE
Begin to build complete Discovery protocol

### PR DESCRIPTION
**Pull request by hayesgm**
This patch begins to build a complete Discovery system. When we join a network, we will use RLPx Discovery Protocol to request peers. For each peer we get, we'll send a ping. If the peer has a reasonable ping time, we'll connect to that peer via the DevP2P TCP protocol. From here, we will begin to require multiple commitments to a block before we sync that block and add it to our chain. Minus a few odds and ends, this is the complete DevP2P experience in Ethereum.

### CODE REVIEW

File: lib/ex_wire/discovery.ex [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    # time. If we like the neighbour, we can try and establish a
+    # RLPx connection.
+    known_nodes = for neighbour <- neighbours, do: neighbour.node
+
+    new_neighbours = Enum.filter(add_neighbours.nodes, fn neighbour ->
+      neighbour.node != ExWire.Config.node_id()
+      and not Enum.member?(known_nodes, neighbour.node)
+    end)
+
+    Logger.debug("[Discovery] Hi-dilly-ho received #{Enum.count(add_neighbours.nodes)} neighboureenos, #{Enum.count(new_neighbours)} newerific")
```
Comment (masonforest) [inline lib/ex_wire/discovery.ex:104]: f339d76f2c5207496dc7d54e69176a7e--ned-flanders-hand-reference

File: lib/ex_wire/struct/endpoint.ex [hunk 2]
```diff
@@ -119,16 +120,29 @@ defmodule ExWire.Struct.Endpoint do
...
+      {p_0, p_1, p_2, p_3} -> <<p_0, p_1, p_2, p_3>>
+      {p_0, p_1, p_2, p_3, p_4, p_5, p_6, p_7} ->
+        encode_word(p_0) <>
+        encode_word(p_1) <>
+        encode_word(p_2) <>
+        encode_word(p_3) <>
+        encode_word(p_4) <>
+        encode_word(p_5) <>
+        encode_word(p_6) <>
+        encode_word(p_7)
```
Comment (masonforest) [inline lib/ex_wire/struct/endpoint.ex:144]: would

```
 Tuple.to_list
  |> :binary.list_to_bin
```
work?

File: lib/ex_wire/config.ex [hunk 3]
```diff
@@ -3,11 +3,15 @@ defmodule ExWire.Config do
   General configuration information for ExWire.
   """
 
-  @port Application.get_env(:ex_wire, :port, 30304)
+  @port Application.get_env(:ex_wire, :port, 30303 + :rand.uniform(10_000))
```
Comment (masonforest) [inline lib/ex_wire/config.ex:6]: Just curious: why do we bind on a random port?

**Review by masonforest — APPROVED** [review]
👍

File: lib/ex_wire/config.ex (hunk 3 above)
Comment (hayesgm) [reply lib/ex_wire/config.ex:6]: If I had to guess, it was for testing and dealing with UDP issues. I'm not sure if there's a good reason to go one way or the other with this.

File: lib/ex_wire/discovery.ex (hunk 1 above)
Comment (hayesgm) [reply lib/ex_wire/discovery.ex:104]: I was going pretty hard insane writing and debugging this code!

File: lib/ex_wire/struct/endpoint.ex (hunk 2 above)
Comment (hayesgm) [reply lib/ex_wire/struct/endpoint.ex:144]: You can check. I think this was just my first foray into ipv6 addresses

**Review by hayesgm — COMMENTED** [review]
Looks good to merge to me.

**Comment by atoulme** [comment]
Besides the conflict, anything blocking this PR to be merged? 
It looks like there is already some discovery code in the repo, is this PR now a duplicate?

**Pull request merged**